### PR TITLE
Change solution to count satellites, not tracking channels

### PR DESCRIPTION
**Pull request by rge-exafore**
Requires https://github.com/swift-nav/libswiftnav/pull/354

### CODE REVIEW

**Comment by rge-exafore** [comment]
@swift-nav/firmware please review

**Comment by ljbade** [comment]
Make sure the correct satellite count is also sent in the SBP solution messages.

File: src/sid_list.c [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <assert.h>
+#include "sid_list.h"
+
+void sid_list_init(gnss_signal_list_t *list, const u16 len, gnss_signal_t *arr)
```
Comment (pmiettinen) [inline src/sid_list.c]: 'len' parameter is probably more suitable named as 'size' seeing how it's used

File: src/sid_list.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#define SRC_SID_LIST_H_
+
+#include <libswiftnav/common.h>
+#include <libswiftnav/signal.h>
+
+typedef struct {
+  gnss_signal_t *l;
+  u16 len;
+  u16 max_len;
+} gnss_signal_list_t;
```
Comment (pmiettinen) [inline src/sid_list.h]: Consider renaming variables, eg.

typedef struct {
   gnss_signal_t *l;
   u16 curr_idx;
   u16 size;
 } gnss_signal_list_t;

Comment (rge-exafore) [inline src/sid_list.h]: len is not an index, it's number of elements. Both 'len' and 'max_len' show number of elements in the list.

File: src/solution.c [hunk 3]
```diff
@@ -420,16 +421,20 @@ static void solution_thread(void *arg)
...
+        if (sid_list_find(&sids_in_track, meas[n_ready].sid) < 0) {
+          sid_list_add(&sids_in_track, meas[n_ready].sid);
+        }
         n_ready++;
       }
       tracking_channel_unlock(i);
     }
 
-    if (n_ready < 4) {
+    if (sids_in_track.len < 4) {
```
Comment (adel-mamin) [inline src/solution.c]: Do I miss a point or the code still counts tracking channels instead of satellites?

File: src/sid_list.c [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  assert (list != NULL);
+  assert (arr != NULL);
+  assert (len > 0);
+
+  list->l = arr;
+  list->max_len = len;
+  list->len = 0;
+}
+
+void sid_list_add(gnss_signal_list_t *list, const gnss_signal_t sid)
```
Comment (adel-mamin) [inline src/sid_list.c]: All sid_list_... functions are missing descriptions.

File: src/solution.c (hunk 3 above)
Comment (pmiettinen) [inline src/solution.c]: n_ready and sids_in_track.len seem to be redundant.

File: src/sid_list.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  sid_list->length = 0;
+}
+
+/** Add new element to the list
+ *
+ * \param sid_list gnss_signal_list_t to add new element to
+ * \param sid element to be added
+ *
+ */
+void sid_list_add(gnss_signal_list_t *sid_list, const gnss_signal_t sid)
```
Comment (adel-mamin) [inline src/sid_list.c]: "const gnss_signal sid": is 'const' really needed here?

**Comment by adel-mamin** [comment]
Looks good to me.

File: src/sid_set.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef SRC_SID_LIST_H_
```
Comment (adel-mamin) [inline src/sid_set.h]: SRC_SID_LIST_H_ -> SRC_SID_SET_H

File: src/sid_set.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <assert.h>
+#include <string.h>
+#include <libswiftnav/bits.h>
+#include "sid_set.h"
+
+/** Initialize new sid list
```
Comment (adel-mamin) [inline src/sid_set.c]: list -> set ?

File: src/sid_set.c [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * \returns Number of unique satellites present in the list.
+ *
+ */
+u32 sid_set_get_sat_count(const gnss_signal_set_t sid_list)
+{
+  u32 cnt = get_set_bit_number(sid_list[CODE_GPS_L1CA] |
+                               sid_list[CODE_GPS_L2CM]);
+
+  cnt += get_set_bit_number(sid_list[CODE_SBAS_L1CA]);
```
Comment (adel-mamin) [inline src/sid_set.c]: Do we really use SBAS SVs in solution?

File: src/sid_set.h (hunk 6 above)
Comment (rge-exafore) [inline src/sid_set.h]: Done.

Comment (rge-exafore) [inline src/sid_set.h]: Done.

File: src/sid_set.c (hunk 8 above)
Comment (rge-exafore) [inline src/sid_set.c]: We may have.

File: src/sid_set.h [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef SRC_SID_SET_H_
+#define SRC_SID_SET_H_
+
+#include <libswiftnav/common.h>
+#include <libswiftnav/signal.h>
+
+typedef u32 gnss_signal_set_t[CODE_COUNT];
```
Comment (jacobmcnamee) [inline src/sid_set.h]: It's generally a bad idea to hide an array behind a typedef. Better to do
```
typedef struct {
  u32 sats;
} gnss_signal_set_t;
```

File: src/sid_set.c [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+/** Add new element to the list
+ *
+ * \param sid_list gnss_signal_list_t to add new element to
+ * \param sid element to be added
+ *
+ */
+void sid_set_add(gnss_signal_set_t sid_list, const gnss_signal_t sid)
+{
+  u32 m = 0x01 << sid_to_code_index(sid);
```
Comment (jacobmcnamee) [inline src/sid_set.c]: Obviously this breaks if `NUM_SATS_<code>`  > 32

File: src/sid_set.c [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+/** Get number of satellites in sid list
+ *
+ * \param sid_list gnss_signal_list_t to add new element to
+ *
+ * \returns Number of unique satellites present in the list.
+ *
+ */
+u32 sid_set_get_sat_count(const gnss_signal_set_t sid_list)
+{
+  u32 cnt = get_bits_number(sid_list[CODE_GPS_L1CA] |
```
Comment (jacobmcnamee) [inline src/sid_set.c]: Can you iterate over the code types programmatically instead of hard-coding them?

File: src/solution.c [hunk 12]
```diff
@@ -420,16 +421,19 @@ static void solution_thread(void *arg)
     gps_time_t rec_time = rx2gpstime(rec_tc);
     u8 n_ready = 0;
     channel_measurement_t meas[MAX_CHANNELS];
+    gnss_signal_set_t sats_in_track;
+    sid_set_init(sats_in_track);
     for (u8 i=0; i<nap_track_n_channels; i++) {
       tracking_channel_lock(i);
       if (use_tracking_channel(i)) {
         tracking_channel_measurement_get(i, rec_tc, &meas[n_ready]);
+        sid_set_add(sats_in_track, meas[n_ready].sid);
```
Comment (jacobmcnamee) [inline src/solution.c]: I'd prefer to move this into a separate loop.

Comment (rge-exafore) [inline src/solution.c]: Done

File: src/sid_set.h (hunk 9 above)
Comment (rge-exafore) [inline src/sid_set.h]: Done

File: src/sid_set.c (hunk 11 above)
Comment (rge-exafore) [inline src/sid_set.c]: I can't as I need to combine data for two GPS and two GLO bands.

File: src/sid_set.c (hunk 10 above)
Comment (rge-exafore) [inline src/sid_set.c]: Changed to 64 bits.

File: src/sid_set.c [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+/** Get number of satellites in sid set
+ *
+ * \param sid_set gnss_sid_set_t to count number of satellites in
+ *
+ * \returns Number of unique satellites present in the set.
+ *
+ */
+u32 sid_set_get_sat_count(const gnss_sid_set_t *sid_set)
+{
+  u32 cnt = count_bits_u64(sid_set->sats[CODE_GPS_L1CA] |
```
Comment (jacobmcnamee) [inline src/sid_set.c]:
```
u32 cnt = 0;
for (enum constellation constellation = 0; 
     constellation < CONSTELLATION_COUNT; constellation++) {
  u64 sats = 0;
  for (enum code code = 0; code < CODE_COUNT; code++) {
    if (code_to_constellation(code) == constellation) {
      sats |= sid_set->sats[code];
    }
  }
  cnt += count_bits_u64(sats);
}
```
or even better
```
u64 sats[CONSTELLATION_COUNT];
memset(sats, 0, sizeof(sats));
for (enum code code = 0; code < CODE_COUNT; code++) {
  sats[code_to_constellation(code)] |= sid_set->sats[code];
}

u32 cnt = 0;
for (enum constellation constellation = 0; 
     constellation < CONSTELLATION_COUNT; constellation++) {
  cnt += count_bits_u64(sats[constellation]);
}
```
That way we don't have to remember to update this function when a new code is added.

Comment (rge-exafore) [inline src/sid_set.c]: I didn't think about this approach. It's better, yes. Done now.

File: src/sid_set.h [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <libswiftnav/common.h>
+#include <libswiftnav/signal.h>
+
+typedef struct {
+  u64 sats[CODE_COUNT];
+} gnss_sid_set_t;
+
+void sid_set_init(gnss_sid_set_t *sid_set);
+void sid_set_add(gnss_sid_set_t *sid_set, const gnss_signal_t sid);
```
Comment (jacobmcnamee) [inline src/sid_set.h]: No need for `const` qualifier on `gnss_signal_t sid`

**Comment by rge-exafore** [comment]
Moved to private (https://github.com/swift-nav/piksi_firmware_private/pull/10) and closed.

**Pull request closed without merging**